src/ripple/app/tests/Flow_test.cpp

```diff
@@ -1271,6 +1271,39 @@ struct Flow_test : public beast::unit_test::suite
             BEAST_EXPECT(offer[sfTakerPays] == USD (495));
         }
     }
+    void testSelfFundedXRPEndpoint (bool consumeOffer)
+    {
+        // Test that the deferred credit table is not bypassed for
+        // XRPEndpointSteps. If the account in the first step is sending XRP and
+        // that account also owns an offer that receives XRP, it should not be
+        // possible for that step to use the XRP received in the offer as part
+        // of the payment.
+        testcase("Self funded XRPEndpoint");
+
+        using namespace jtx;
+
+        Env env(*this, features(featureFlowV2));
+
+        // Need new behavior from `accountHolds`
+        auto const closeTime = amendmentRIPD1141SoTime() +
+            env.closed()->info().closeTimeResolution;
+        env.close(closeTime);
+
+        auto const alice = Account("alice");
+        auto const gw = Account("gw");
+        auto const USD = gw["USD"];
+
+        env.fund(XRP(10000), alice, gw);
+        env(trust(alice, USD(20)));
+        env(pay(gw, alice, USD(10)));
+        env(offer(alice, XRP(50000), USD(10)));
+
+        // Consuming the offer changes the owner count, which could also cause
+        // liquidity to decrease in the forward pass
+        auto const toSend = consumeOffer ? USD(10) : USD(9);
+        env(pay(alice, alice, toSend), path(~USD), sendmax(XRP(20000)),
+            txflags(tfPartialPayment | tfNoRippleDirect));
```

Comment:
@seelabs so if `consumeOffer` is true (i.e., `toSend == USD(10)`) the payment would fail because internally the `ownerCount` would be wrong and the forward pass would fail.  I think this comment change might help, but see what you think:
```
// Consuming the offer changes the owner count, which could also cause
// liquidity to decrease in the forward pass.  If the forward pass
// liquidity decreases (because the ownerCount is mismanaged) then
// the payment will fail.
```